Sources/SyncDataProviders/Bookmarks/BookmarksProvider.swift

```diff
@@ -0,0 +1,258 @@
+                        if (error as NSError).code == NSManagedObjectMergeError {
+                            context.reset()
+                        } else {
+                            saveError = error
+                            break
+                        }
+                    }
+
+                }
+            }
+            if let saveError {
+                continuation.resume(throwing: saveError)
+                return
+            }
+
+            if let serverTimestamp {
+                lastSyncTimestamp = serverTimestamp
+                reloadBookmarksAfterSync()
+            }
+
+            continuation.resume(returning: ())
+        }
+    }
+
+    public func handleSyncError(_ error: Error) {
+        syncErrorSubject.send(error)
+    }
+
+    // MARK: - Internal
+
+    func cleanUpSentItems(_ sent: [Syncable], receivedUUIDs: Set<String>, clientTimestamp: Date, in context: NSManagedObjectContext) -> Set<String> {
+        if sent.isEmpty {
+            return []
+        }
+        let identifiers = sent.compactMap(\.uuid)
+        let bookmarks = BookmarkEntity.fetchBookmarks(with: identifiers, in: context)
+
+        var idsOfItemsThatRetainModifiedAt = Set<String>()
+
+        for bookmark in bookmarks {
+            if let modifiedAt = bookmark.modifiedAt, modifiedAt > clientTimestamp {
+                if let uuid = bookmark.uuid {
+                    idsOfItemsThatRetainModifiedAt.insert(uuid)
+                }
+                continue
+            }
+            let isLocalChangeRejectedBySync: Bool = bookmark.uuid.flatMap { receivedUUIDs.contains($0) } == true
+            if bookmark.isPendingDeletion, !isLocalChangeRejectedBySync {
+                context.delete(bookmark)
+            } else {
+                bookmark.modifiedAt = nil
+            }
+        }
+
+        return idsOfItemsThatRetainModifiedAt
+    }
+
+    /**
+     * Saves context and ensures that `modifiedAt` field is cleared on affected objects.
+     *
+     * Context needs to be saved twice because setting `modifiedAt` to `nil` is not enough when `modifiedAt` was already `nil`.
+     * In that case it would get updated to current date. So we first trigger save to populate `modifiedAt`,
+     * and then we explicitly clear them for all affected objects and save again.
+     */
+    private func saveContextAndClearModifiedAt(_ context: NSManagedObjectContext, excludedUUIDs: Set<String> = []) throws {
+        let insertedObjects = Array(context.insertedObjects).compactMap { $0 as? BookmarkEntity }
+        let updatedObjects = Array(context.updatedObjects.subtracting(context.deletedObjects)).compactMap { $0 as? BookmarkEntity }
+
+        try context.save()
+        (insertedObjects + updatedObjects).forEach { bookmarkEntity in
+            if let uuid = bookmarkEntity.uuid, !excludedUUIDs.contains(uuid) {
+                bookmarkEntity.modifiedAt = nil
+            }
+        }
+        try context.save()
```

Comment:
We rather want to avoid double-saving as it could lead to weird issues in case latter save fails.

Let's do what we talked about: set modifiedAt to nil only for known (and up-front specified) set of objects based on a union of sent and received ones. Then filter out these with modifiedAt newer than sync stat date, and just clean that timestamp for the ones that are left.

As for how to mark these object, I think it is acceptable to have in-memory flag associated with manage object entity that is set before save and reset afterwards.

Reply:
I’ll try adding more unit tests to cover various scenarios.